### PR TITLE
Upgrade to RN v0.64!

### PR DESCRIPTION
**Pull request by chrisbobbe**
Following the now-merged PRs #4433, #4518, #4520, #4836, #4843, #4846, #4887, #4894, #4891, #4907, #4928, #4982, #4983, #4985, #4987, and #4989, complete the RN v0.64 upgrade! Some of those were small, some very large. And there were probably other ones, possibly including some we didn't realize were on the path to this upgrade. 🙂

Fixes: #4426

### CODE REVIEW

**Comment by chrisbobbe** [comment]
These are the RN commits that touch the [RN template app](https://github.com/facebook/react-native/tree/master/template) that are in v0.64.2 but are not in v0.63.4. They're ordered from first to last (oldest to newest).

I.e., they're the commits in the output of `git log --oneline --reverse v0.63.4..v0.64.2 -- template`, when I run it in a clone of the facebook/react-native repo.

Along with whether, how, and when we propagate the changes to our app.

We shouldn't forget to update this, if necessary, to follow any changes made during code review and merge.

| RN commit | When done | Comment |
| --- | --- | --- |
| facebook/react-native@cd13c99d0 Update default Podfile to not depend on a path (#28572) | 0a2a6ac1f | We should have taken facebook/react-native@696fb5588 in the RN v0.63 upgrade (#4420), but forgot. Oops. |
| facebook/react-native@8988a073b Update Gradle Wrapper to 6.3 (#28173) | cb316419f | We went all the way to 6.7.1. |
| facebook/react-native@33b3a1a14 Changed iOS LaunchScreen from xib to storyboard (#28239) | 78a62b249 | … |
| facebook/react-native@44beb2a68 Remove the post install step (#28651) | N/A | Cleanly reverted in the next commit, facebook/react-native@fb04237b1, so, ignore. |
| facebook/react-native@fb04237b1 Revert D21064653: Remove the post install step | N/A | Cleanly reverts the previous commit, facebook/react-native@44beb2a68, so, ignore. |
| facebook/react-native@17f025bc2 Upgrade Flipper to 0.37.0 (#28545) |  f7005c50c and 271d9c482 | … |
| facebook/react-native@57fee3389 Flow 0.123.0 in xplat/js | 6c1e4b757 | … |
| facebook/react-native@df03228a6 remove tvOS from template (#28706) | N/A | We didn't seem to have any of the things that this commit removes. We've already trimmed away (or we never had) this tvOS-specific stuff. |
| facebook/react-native@2b56011f9 Add Dark Mode support to the App template and NewAppScreen components (#28711) | 4db472a2e | We don't take all of the changes because of some existing customizations. |
| facebook/react-native@d8e6c4578 Move DevSettingsActivity from main to debug (#28770) | fc17b21e0 | … |
| facebook/react-native@780f06cd4 Bump @react-native-community/eslint-config in new app template | N/A | We don't use @react-native-community/eslint-config. We might eventually start to; see #4119. |
| facebook/react-native@ced959bb3 chore: remove Kotlin version from the default template (#28626) | N/A | We should leave these lines in because we use Kotlin; see [discussion](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Kotlin.20in.20RN.20v0.2E63.20upgrade/near/1073269). |
| facebook/react-native@0c2c4bf33 Deploy Flow v0.124.0 to xplat/js | 6c1e4b757 | … |
| facebook/react-native@7bb1c4e1b Enable with CocoaPods `:configuration` (#28796) | 1ade53db9 | It looks like this was backported to RN v0.63 (facebook/react-native@29639e7b9), and we handled it when we were doing that upgrade. |
| facebook/react-native@7b61a968f set`NSAllowsArbitraryLoads` to false by default in template (#28885) | N/A | We didn't have `NSAllowsArbitraryLoads` set to `true`. We mention how to set this setting in our docs/howto/ios-tips.md. |
| facebook/react-native@37e7b4141 deploy Flow 0.125.1 to xplat | c4d2a37ff | … |
| facebook/react-native@afad239d7 remove suppress_comments from xplat flowconfigs | c666dafd9 | … |
| facebook/react-native@3b96b39ab Deploy Flow 0.126.1 | 03e6bdffb | … |
| facebook/react-native@9a532edaa upgrade to flow v0.127.0 | 03e6bdffb | … |
| facebook/react-native@d8af579f5 Upgrade dependency Metro bundler to v0.60.0 | 77ba1726a | … |
| facebook/react-native@a36d9cd7e Deploy 0.128 to xplat/js | f404f4202 | … |
| facebook/react-native@e559aee64 bump gradle to 6.5 (#29043) | cb316419f | … |
| facebook/react-native@a5151c2b5 Deploy 0.129 to xplat/js | This PR (main upgrade commit) | … |
| facebook/react-native@1c634a921 move Hermes badge from template to NewAppScreen library (#28783) | N/A | Refactoring of template's App.js; we don't need to follow this. |
| facebook/react-native@7694b32a8 Exclude okhttp from flipper dependency (#29260) | 48b223729 | … |
| facebook/react-native@e629e94b4 Update Android build tools to 29.0.3 (#29350) | N/A | We don't need to specify `buildToolsVersion`; see e0b0b97d2 and [discussion](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/buildToolsVersion.20in.20RN.20v0.2E63.20upgrade/near/1073261). |
| facebook/react-native@b98c6f208 Revert React version back to 0.0.0-experimental-aae83a4b9 | N/A | In package.json, "0.0.0-experimental-aae83a4b9" will not be correct for `dependencies.react` or `devDependencies.react-test-renderer`. |
| facebook/react-native@da865b23a Update Metro packages to v0.61.0 | 77ba1726a | … |
| facebook/react-native@050a7dd01 Enable exact objects by default in the project template | This PR (main upgrade commit) | This took a lot of work. Glad to have it done! |
| facebook/react-native@7204a58c1 Deploy 0.130.0 to Xplat | This PR (main upgrade commit) | … |
| facebook/react-native@79d0a7d71 Upgrade minsdkversion of RN OSS template to API level 19 | N/A | We're already on 21. |
| facebook/react-native@b6ba6f580 "React Native sync for revisions b5c6dd2...a437f3ff3" | N/A | Cleanly reverted in the next commit, facebook/react-native@4cdb1951b, so, ignore. |
| facebook/react-native@4cdb1951b Back out ""React Native sync for revisions b5c6dd2...a437f3ff3"" | N/A | Cleanly reverts the previous commit, facebook/react-native@b6ba6f580, so, ignore. |
| facebook/react-native@d0e710bfb Re-land React Native sync for revisions b5c6dd2...e67a6b160 | N/A | Cleanly reverted in facebook/react-native@499b021b7, so, ignore. |
| facebook/react-native@471c2ebe5 Upgrade to Flow v0.131.0 | This PR (main upgrade commit) | … |
| facebook/react-native@499b021b7 Backout React Sync again | N/A | Cleanly reverts facebook/react-native@d0e710bfb, so, ignore. |
| facebook/react-native@b41f6875e Update Metro packages to v0.62.0 | 77ba1726a | We're already on v0.63. |
| facebook/react-native@abba44de1 Re-land React Sync b5c6dd2...a437f3ff3 | This PR (main upgrade commit) | This just bumps versions of "react" and "react-test-renderer". We'll take React 17 in the main commit, and we're not currently using "react-test-renderer". |
| facebook/react-native@fcffb961f remove redundant deprecated-utility setting from flow configs | ec10d120d | … |
| facebook/react-native@464293f64 Update Metro packages to v0.63.0 | 77ba1726a | We're already on v0.63. |
| facebook/react-native@4aa69e634 0.132 for xplat | This PR (main upgrade commit) | … |
| facebook/react-native@959365a90 Change default metro.config.js to utilize inline requires (#29770) | 7234ab175 | … |
| facebook/react-native@5bc67b658 Update Gradle Wrapper to 6.6 (#29613) | cb316419f | We're already on 6.7. |
| facebook/react-native@cf8368f20 Update Android Gradle plugin to 3.6.4 (#28170) | This PR | … |
| facebook/react-native@54e19a6b7 RN: Remove `fbjs` Dependency | ed6deb523 | … |
| facebook/react-native@3a8559b86 Use Gradle Wrapper 6.6 (-all variant) (#29793) | N/A | We're already using the "-all" variant, intentionally (see a1c9f1b77). |
| facebook/react-native@fc2b1cac8 Deploy Flow v0.133.0 | This PR (main upgrade commit) | … |
| facebook/react-native@fb354cba0 Remove .gitattributes and normalize line endings (#29792) | N/A | We're happy with our .gitattributes. |
| facebook/react-native@635ac1ba5 Update various Xplat Prettier configs to specify arrowParens: avoid | 7e4f7a405 | … |
| facebook/react-native@18ffe1220 Add an explicit NDK version to Android template (#29403) | N/A | https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Pin.20to.20explicit.20NDK.20version.3F/near/1131628 |
| facebook/react-native@d8b70b19b Update Flipper (#29787) | 0dfc9af7f | … |
| facebook/react-native@54016faa7 Deploy Flow v0.134.0 | This PR (main upgrade commit) | … |
| facebook/react-native@e1bf515ae Making Android versionCodeOverride for new apps using the template human-readable (#29808) | eec8c05be | … |
| facebook/react-native@e599d6c5d Fix passing react native path in Podfile template (#29285) | 0a2a6ac1f | … |
| facebook/react-native@216037757 remove most of tvOS remnants from the code (#29407) | 64e5f8c12 | … |
| facebook/react-native@553fb8b28 Update Android Gradle plugin to 4.0.1 (#29013) | This PR | Issue with this version identified at https://github.com/facebook/react-native/pull/29013#issuecomment-708079213. A fix was supposedly done in a RN internal change in facebook/react-native@53f55001a, released in RN v0.64. So, doing this after taking RN v0.64. |
| facebook/react-native@dd35bf534 Deploy Flow v0.135.0 | This PR (main upgrade commit) | … |
| facebook/react-native@6685aba46 Update template to xcode 12 (#30150) | ed91c3d80 | We don't make all these changes; see [discussion](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/RN.20v0.2E64.20upgrade.3A.20Xcode.20config.20changes.3F/near/1209893). |
| facebook/react-native@500bdf74d Deploy Flow v0.136.0 | This PR (main upgrade commit) | … |
| facebook/react-native@d03c0f965 Generate FBReactNativeSpec ObjC++ source files using CocoaPods | N/A | The Podfile change gets cleanly reverted in facebook/react-native@d8a0a9b1d; the added dependency is removed in facebook/react-native@14db55696. So, don't take this. |
| facebook/react-native@a17ff44ad Upgrade minsdkversion of RN OSS template to API level 21 | N/A | Ours is already at 21. |
| facebook/react-native@24bca492c Update React Native to React 17.0.0 | This PR (main upgrade commit) | … |
| facebook/react-native@dfa9db49e bump Android Gradle Plugin to 4.1.0 (#30201) | This PR | See note on handling facebook/react-native@553fb8b28; we do this after taking RN v0.64. |
| facebook/react-native@e9fd93f53 Bump Gradle Wrapper to 6.7 (#30200) | N/A | We're already on 6.7.1, which is the latest 6.7.x: https://gradle.org/releases/ |
| facebook/react-native@be8fe21d0 Deploy Flow v0.137.0 | This PR (main upgrade commit) | … |
| facebook/react-native@538446e75 Bump Xcode to 12.1.0, bump CocoaPods 1.10.0 (#30250) | 5bd6cd5ec | … |
| facebook/react-native@06fedcd32 Upgrade metro to 0.64.0 | cabab2243 | … |
| facebook/react-native@68110d4ab Use codegen from source in default iOS template apps | N/A | This was part of work started in facebook/react-native@d03c0f965 and ultimately reverted in facebook/react-native@d8a0a9b1d and facebook/react-native@14db55696. |
| facebook/react-native@65003d743 Use react-native-codegen@0.0.6 in new app template | N/A | This was part of work started in facebook/react-native@d03c0f965 and ultimately reverted in facebook/react-native@d8a0a9b1d and facebook/react-native@14db55696. |
| facebook/react-native@0117077b9 [0.64.0-rc.0] Bump version numbers | This PR (main upgrade commit) | … |
| facebook/react-native@d197e9589 [0.64.0-rc.1] Bump version numbers | This PR (main upgrade commit) | … |
| facebook/react-native@d8a0a9b1d Integrate Native Module codegen into Xcode build pipeline (#30449) | N/A | Cleanly reverts the Podfile change from facebook/react-native@d03c0f965, which we didn't take. |
| facebook/react-native@fffa4d1ae Update template devDependencies (#30489) | ce11ee6af | … |
| facebook/react-native@eb85d1dbd Add instructions to template/ios/Podfile for enabling hermes (#30461) | This PR | Done, with facebook/react-native@2ba3ef5b1 squashed in. |
| facebook/react-native@2ba3ef5b1 fix: default template on iOS (#30571) | This PR | Squashed in with the commit that took facebook/react-native@eb85d1dbd. |
| facebook/react-native@5c09b3fb5 fix: building in release mode for simulator (#30543) | This PR | This depends on a script that ships with RN; so, we take this after taking RN v0.64. |
| facebook/react-native@ed237b479 Exclude `i386` from valid architectures when building with Hermes on iOS (#30592) | This PR | This depends on a script that ships with RN; so, we take this after taking RN v0.64. |
| facebook/react-native@1eb7d4aef [0.64.0-rc.2] Bump version numbers | This PR (main upgrade commit) | … |
| facebook/react-native@c023a4061 chore: bump codegen script | N/A | This was part of work started in facebook/react-native@d03c0f965 and ultimately reverted in facebook/react-native@d8a0a9b1d and facebook/react-native@14db55696. |
| facebook/react-native@e846740d3 [0.64.0-rc.3] Bump version numbers | This PR (main upgrade commit) | … |
| facebook/react-native@7159bcbf7 Update flipper in RNTester and template (#31010) | aa70f106f | … |
| facebook/react-native@14db55696 fix: React Native CodeGen integration for 0.64-stable (#31027) | N/A | Removes the dependency that would've been added in facebook/react-native@d03c0f965, which we didn't take. |
| facebook/react-native@1aa4f47e2 [0.64.0-rc.4] Bump version numbers | This PR (main upgrade commit) | … |
| facebook/react-native@728d55a3e Fixing the git attrs for all the people and all the files and all future 🙌 | N/A | This seems to address some problems introduced in facebook/react-native@fb354cba0, which we decided not to take (see above). See [discussion](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/RN.20v0.2E64.20upgrade.3A.20.2Egitattributes.20and.20.2Eeditorconfig.3F/near/1222700) |
| facebook/react-native@ace025d2c [0.64.0] Bump version numbers | This PR (main upgrade commit) | … |
| facebook/react-native@787567a15 [0.64.1] Bump version numbers | This PR (main upgrade commit) | … |
| facebook/react-native@757bb75fb [0.64.2] Bump version numbers | This PR (main upgrade commit) | … |

**Review by gnprice — COMMENTED** [review]
Thanks @chrisbobbe! There was a lot in this upgrade project; excited to see it near completion.

I've just read up through the main upgrade commit, and have only the comment below. I'll plan to look at the rest tomorrow when we're both in the office.

File: jest.config.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ const transformModulesWhitelist = [
   'expo-application',
   'expo-web-browser',
   'react-native',
+  '@react-native',
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply jest.config.js:10]: Yeah, I'm not totally sure, but I think it's likely that we can't easily extend one or both. `transformIgnorePatterns` is an array, so we could potentially just concat our own custom thing onto one or both of the others. But I think once an item in that array says to ignore a path, you probably can't have a different item cause it to be un-ignored. And each of these items expresses what paths it wants compiled by saying "ignore everything except for this narrow range" and those ranges don't overlap.

[the remark this answers is not in the document]
Comment (gnprice) [reply jest.config.js:10]: Cool, yeah, that makes sense.

In that case I'd just tweak this message so it's not pointing toward "alas, we don't extend those and instead have chosen to ignore and clobber them". Mentioning that this item appears in those is fine context to have, but this API just isn't suitable to taking a list from elsewhere and extending it.

**Comment by gnprice** [comment]
OK, finished reading the rest of the branch, and all looks good! One commit-message quibble in the thread above; otherwise, please merge at will :rocket:

**Comment by chrisbobbe** [comment]
> OK, finished reading the rest of the branch, and all looks good! One commit-message quibble in the thread above; otherwise, please merge at will 🚀

Thanks for the review! Merged.

**Pull request merged**